sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,968 @@
+     * for further use by splitIntoBundles() and getEstimatesSizeBytes().
+     * @throws InterruptedException
+     */
+    @VisibleForTesting
+    void computeSplits() throws IOException, InterruptedException{
+      inputFormatObj = createInputFormat();
+      List<InputSplit> splits =
+          inputFormatObj.getSplits(Job.getInstance(conf.getHadoopConfiguration()));
+      if (splits == null) {
+        throw new IOException(HadoopInputFormatIOContants.COMPUTESPLITS_NULL_GETSPLITS_ERROR_MSG);
+      }
+      if (splits.isEmpty()) {
+        throw new IOException(HadoopInputFormatIOContants.COMPUTESPLITS_EMPTY_SPLITS_ERROR_MSG);
+      }
+      boundedSourceEstimatedSize = 0;
+      inputSplits = new ArrayList<SerializableSplit>();
+      for (InputSplit inputSplit : splits) {
+        if (inputSplit == null) {
+          throw new IOException(HadoopInputFormatIOContants.COMPUTESPLITS_NULL_SPLIT_ERROR_MSG);
+        }
+        boundedSourceEstimatedSize = boundedSourceEstimatedSize + inputSplit.getLength();
+        inputSplits.add(new SerializableSplit(inputSplit));
+      }
+      validateKeyValueClasses();
+    }
+
+    /**
+     * Returns instance of InputFormat set in the configuration.
+     */
+    private InputFormat<?, ?> createInputFormat() throws IOException, InterruptedException {
+      InputFormat<?, ?> inputFormatObj;
+      try {
+        inputFormatObj = (InputFormat<?, ?>) conf.getHadoopConfiguration()
+            .getClass(HadoopInputFormatIOContants.INPUTFORMAT_CLASSNAME, null).newInstance();
+        /**
+         * Sets configuration if provided InputFormat implements Configurable.
+         */
+        if (Configurable.class.isAssignableFrom(inputFormatObj.getClass())) {
+          ((Configurable) inputFormatObj).setConf(conf.getHadoopConfiguration());
+        }
+      } catch (InstantiationException e) {
+        throw new IOException("Unable to create InputFormat : ", e);
+      } catch (IllegalAccessException e) {
+        throw new IOException("Unable to create InputFormat : ", e);
+      }
+      return inputFormatObj;
+    }
+
+    /**
+     * Throws exception is InputFormat key/value class set by the user is different than InputFormat's actual key/value class.
+     */
+    private void validateKeyValueClasses() throws IOException, InterruptedException {
+      RecordReader<?, ?> reader = inputFormatObj.createRecordReader(inputSplits.get(0).getSplit(),
+          new TaskAttemptContextImpl(conf.getHadoopConfiguration(), new TaskAttemptID()));
+      if (reader == null) {
+        throw new IOException(
+            String.format(HadoopInputFormatIOContants.NULL_CREATE_RECORDREADER_ERROR_MSG,
+                inputFormatObj.getClass()));
+      }
+      reader.initialize(inputSplits.get(0).getSplit(),
+          new TaskAttemptContextImpl(conf.getHadoopConfiguration(), new TaskAttemptID()));
+      reader.nextKeyValue();
```

Comment:
Yes, to get the actual key and value class of InputFormat for validation. Added a comment accordingly "// First record is read to get the InputFormat's key and value classes. "